### PR TITLE
Increase AsyncLocal testing

### PR DESCRIPTION
**Pull request by benaadams**
Execution context with locals to and from empty context 
Execution context with locals to and from context with different locals
Test different counts of asynclocals (as types change based on number of locals)

PTAL @stephentoub

### CODE REVIEW

**Comment by benaadams** [comment]
test Linux x64 Release Build please

**Comment by benaadams** [comment]
```
opensuse.422.amd64.open-Release-x64
System.Net.Http.Functional.Tests
 System.Net.Http.Functional.Tests.HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test
/SetDelegate_ConnectionSucceeds(acceptedProtocol: Tls, requestOnlyThisProtocol: False)

Unhandled Exception of Type System.Net.Http.HttpRequestException
Message :
System.Net.Http.HttpRequestException : An error occurred while sending the request.
---- System.Net.Http.CurlException : Server returned nothing (no headers, no data)
Stack Trace :
   at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__62.MoveNext() in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/HttpClient.cs:line 488
   at System.Net.Http.Functional.Tests.HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test.<>c__DisplayClass3_1.<<SetDelegate_ConnectionSucceeds>b__0>d.MoveNext() in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.AcceptAllCerts.cs:line 53
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Test.Common.LoopbackServer.<>c__DisplayClass3_0.<CreateServerAsync>b__0(Task t) in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/Common/tests/System/Net/Http/LoopbackServer.cs:line 68
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state) in /root/coreclr/src/mscorlib/shared/System/Threading/ExecutionContext.cs:line 151
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot) in /root/coreclr/src/mscorlib/src/System/Threading/Tasks/Task.cs:line 2440
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Http.Functional.Tests.HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test.<SetDelegate_ConnectionSucceeds>d__3.MoveNext() in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.AcceptAllCerts.cs:line 51
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace -----
   at System.Net.Http.CurlHandler.ThrowIfCURLEError(CURLcode error) in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs:line 663
   at System.Net.Http.CurlHandler.MultiAgent.FinishRequest(StrongToWeakReference`1 easyWrapper, CURLcode messageResult) in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:line 883
```

**Comment by benaadams** [comment]
test Linux x64 Release Build please

**Comment by benaadams** [comment]
Fedora 
System.Net.Sockets.Tests
```
Discovering: System.Net.Sockets.Tests
Discovered:  System.Net.Sockets.Tests
Starting:    System.Net.Sockets.Tests

Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Net.Sockets.SocketAsyncContext.AcceptOperation.InvokeCallback() in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:line 543
   at System.Net.Sockets.SocketAsyncContext.OperationQueue`1.ProcessQueue(SocketAsyncContext context) in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:line 947
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state) in /root/coreclr/src/mscorlib/shared/System/Threading/ExecutionContext.cs:line 151
   at System.Threading.ThreadPoolWorkQueue.Dispatch() in /root/coreclr/src/mscorlib/src/System/Threading/ThreadPool.cs:line 588
/home/helixbot/dotnetbuild/work/1627409d-2dd2-4c0e-a4eb-23635663fba5/Work/a6e0be47-afcd-44ee-889f-115d0e29eb8a/Unzip/RunTests.sh: line 87: 129646 Aborted                 (core dumped) $RUNTIME_PATH/dotnet xunit.console.netcore.exe System.Net.Sockets.Tests.dll -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapptests -notrait category=nonlinuxtests -notrait category=OuterLoop -notrait category=failing
Trying to find crash dumps for project: System.Net.Sockets.Tests
No new dump file was found in /home/helixbot/dotnetbuild/work/1627409d-2dd2-4c0e-a4eb-23635663fba5/Work/a6e0be47-afcd-44ee-889f-115d0e29eb8a/Unzip
```
NullReferenceException, exotic

**Comment by benaadams** [comment]
test Linux x64 Release Build please

File: src/System.Threading/tests/AsyncLocalTests.cs [hunk 1]
```diff
@@ -419,5 +423,162 @@ private static Task Run(Func<Task> func)
...
+
+        public static IEnumerable<object[]> GetCount()
+        {
+            const int max = 40;
+
+            foreach (int i in Enumerable.Range(1, max))
+            {
+                yield return new object[] { i };
+            }
+        }
```
Comment (stephentoub) [inline src/System.Threading/tests/AsyncLocalTests.cs]: Might be simpler with LINQ, e.g.
```C#
public static IEnumerable<object[]> GetCounts() =>
    Enumerable.Range(0, 40).Select(i => new object[] { i });
```

File: src/System.Threading/tests/AsyncLocalTests.cs [hunk 2]
```diff
@@ -419,5 +423,162 @@ private static Task Run(Func<Task> func)
...
+
+            ExecutionContext.SuppressFlow();
+            try
+            {
+                TestExecutionContextRun();
+            }
+            finally
+            {
+                ExecutionContext.RestoreFlow();
+            }
```
Comment (stephentoub) [inline src/System.Threading/tests/AsyncLocalTests.cs]: There's a lot happening in this one test; can it be split apart?

File: src/System.Threading/tests/AsyncLocalTests.cs [hunk 3]
```diff
@@ -419,5 +423,162 @@ private static Task Run(Func<Task> func)
...
+                        for (int l = 0; l < locals.Length; l++)
+                        {
+                            Assert.Equal(0, setManually[l]);
+                            Assert.Equal(l < i ? 0 : 1, unsetAutomatically[l]);
+                            Assert.Equal(l < i ? 0 : 2, setAutomatically[l]);
+                        }
+
+                        for (int c = 0; c < locals.Length; c++)
+                        {
+                            ExecutionContext.Run(contexts[c].CreateCopy(), _ =>
```
Comment (stephentoub) [inline src/System.Threading/tests/AsyncLocalTests.cs]: It's possible I missed it, but do you explicitly test setting values within ExecutionContext.Run and making sure they've not propagated out?  That could/should be a separate test as well.

File: src/System.Threading/tests/AsyncLocalTests.cs [hunk 4]
```diff
@@ -419,5 +423,162 @@ private static Task Run(Func<Task> func)
...
+
+            await AsyncRecursive(count - 1, locals, contexts, args =>
+            {
+                if (!args.ThreadContextChanged)
+                {
+                    if (args.PreviousValue != null)
+                    {
+                        Assert.True(false, $"Trying to set {args.CurrentValue} when {args.PreviousValue} value already set");
+                    }
+
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/System.Threading/tests/AsyncLocalTests.cs]: Done

File: src/System.Threading/tests/AsyncLocalTests.cs (hunk 1 above)
Comment (benaadams) [reply src/System.Threading/tests/AsyncLocalTests.cs]: Done, though starting from 1 (for 1 asynclocal)

File: src/System.Threading/tests/AsyncLocalTests.cs [hunk 5]
```diff
@@ -419,5 +423,162 @@ private static Task Run(Func<Task> func)
...
+            if (index > 0)
+            {
+                await AsyncRecursive(index - 1, locals, contexts, valueChangedHandler);
+            }
+        }
+
+        public static IEnumerable<object[]> GetCount()
+        {
+            const int max = 40;
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/System.Threading/tests/AsyncLocalTests.cs]: Done

File: src/System.Threading/tests/AsyncLocalTests.cs [hunk 6]
```diff
@@ -419,5 +423,162 @@ private static Task Run(Func<Task> func)
...
+                    }, null);
+
+                    Array.Clear(setManually, 0, count);
+                    Array.Clear(unsetAutomatically, 0, count);
+                    Array.Clear(setAutomatically, 0, count);
+                }
+            }
+        }
+
+        static async Task AsyncRecursive(int index, AsyncLocal<object>[] locals, ExecutionContext[] contexts, Action<AsyncLocalValueChangedArgs<object>> valueChangedHandler)
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/System.Threading/tests/AsyncLocalTests.cs]: Done, and made local

File: src/System.Threading/tests/AsyncLocalTests.cs (hunk 3 above)
Comment (benaadams) [reply src/System.Threading/tests/AsyncLocalTests.cs]: Added

File: src/System.Threading/tests/AsyncLocalTests.cs (hunk 2 above)
Comment (benaadams) [reply src/System.Threading/tests/AsyncLocalTests.cs]: Refactored into locals and added comments; would be a lot of setup repetition if I split the test; or lots of params for functions.

Better now though; or should I split?

File: src/System.Threading/tests/AsyncLocalTests.cs [hunk 7]
```diff
@@ -419,5 +472,180 @@ private static Task Run(Func<Task> func)
...
+            ValidateAsyncLocalsValuesNull();
+
+            // Check Running with the contexts captured when setting the locals
+            TestCapturedExecutionContexts();
+
+            ExecutionContext.SuppressFlow();
+            try
+            {
+                // Re-check restoring, but starting with a suppressed flow
+                TestCapturedExecutionContexts();
```
Comment (benaadams) [inline src/System.Threading/tests/AsyncLocalTests.cs]: This caught me with a null ref in "Reduce Execution Context Save+Restore" fix https://github.com/dotnet/coreclr/pull/15629/commits/c11e2e1736e28c6d7b57165a065d2f109b193dc8

In a test that wasn't for async locals

**Comment by tarekgh** [comment]
@benaadams did you run this test against netfx too?

**Comment by benaadams** [comment]
Yep
```
RunTestsForProject:
  C:\GitHub\corefx\bin/tests/System.Threading.Tests/netfx-Windows_NT-Debug-x64//RunTests.cmd C:\GitHub\corefx\bin/testhost/netfx-Windows_NT-Debug-x64/
  Using C:\GitHub\corefx\bin\testhost\netfx-Windows_NT-Debug-x64\ as the test runtime folder.
  Executing in C:\GitHub\corefx\bin\tests\System.Threading.Tests\netfx-Windows_NT-Debug-x64\
  Running tests... Start time: 19:32:06.10
  xUnit.net Console Runner (64-bit Desktop .NET 4.0.30319.42000)
    Discovering: System.Threading.Tests
    Discovered:  System.Threading.Tests
    Starting:    System.Threading.Tests
      System.Threading.Tests.MutexTests.Ctor_TryCreateGlobalMutexTest_Uwp [SKIP]
        Condition(s) not met: "IsInAppContainer"
    Finished:    System.Threading.Tests
  === TEST EXECUTION SUMMARY ===
     System.Threading.Tests  Total: 353, Errors: 0, Failed: 0, Skipped: 1, Time: 6.783s
  Finished running tests.  End time=19:32:13.60, Exit code = 0
  Creating "C:\GitHub\corefx\bin/tests/System.Threading.Tests/netfx-Windows_NT-Debug-x64//tests.passed.without.OuterLoop.failing" because "AlwaysCreate" was specified.
```

CI should have run it too?

**Comment by tarekgh** [comment]
@stephentoub do you have more comments? can we merge it now?

**Comment by tarekgh** [comment]
@stephentoub I am merging this and feel free to send more comments if you need to change anything.

Thanks @benaadams for adding this test.

**Pull request merged**